Add gain circle

keikawa opened the pull request:
This is related to #362 and #943.

Vinc0110 commented on skrf/network.py:
```diff
@@ -4381,6 +4381,120 @@ def stability_circle(self, target_port: str, npoints: int = 181) -> npy.ndarray:
+
+        """
+
+        if self.nports != 2:
```
It good to check this. Maybe also add a comment in the docstring on this requirement for a 2-port network.

keikawa replied:
I added comment in the docstring.

Vinc0110 commented on skrf/network.py:
```diff
@@ -4381,6 +4381,120 @@ def stability_circle(self, target_port: str, npoints: int = 181) -> npy.ndarray:
+        >>> rf.plotting.plot_smith(s=sgc, smith_r=1)
+        >>> plt.show()
+
+        References
```
Providing references is very good. But they are not referenced in any of the docstring sections above, are they? (Not sure if they have to be referenced...)

keikawa replied:
I added reference numbers in the docstring section.

Vinc0110 commented on skrf/network.py:
```diff
@@ -4381,6 +4381,120 @@ def stability_circle(self, target_port: str, npoints: int = 181) -> npy.ndarray:
+
+        gain_factor = mf.db10_2_mag(gain) * (1 - npy.abs(reflection) ** 2)
+        if gain_factor > 1:
+            warnings.warn("The specified gain is too large to be realized. Please specify a smaller gain.", RuntimeWarning)
```
I found it helpful to add `stacklevel=2` as an argument for `warnings.warn()`:
https://docs.python.org/3/library/warnings.html#warnings.warn

keikawa replied:
I added `stacklevel=2` argument.

Vinc0110 reviewed (approved):
Very nice! Thanks for adding this PR to quickly!

I like the extensive docstrings and the tests. Most of my comments are very minor. The only thing I don't like is the definition of `target_port` with 'load' and 'source'. Any other opinions on this?

keikawa commented:
Thank you for your kind review and comment.
As you suggested, I fixed the specification of `target_port` and some documentation. 
Additionally, I corrected a bug where the if statement would not execute when `gain_factor` is `ndarray`.

keikawa commented on skrf/network.py:
```diff
@@ -4283,7 +4283,7 @@ def vswr_active(self, a: npy.ndarray) -> npy.ndarray:
     def stability_circle(self, target_port: str, npoints: int = 181) -> npy.ndarray:
         r"""
-        Returns a complex number of stability circles for a given port ('load' or 'source').
+        Returns stability circles for a given port ('load' or 'source').
```
I changed the specification of `target_port`.

keikawa commented on skrf/network.py:
```diff
@@ -4381,6 +4381,120 @@ def stability_circle(self, target_port: str, npoints: int = 181) -> npy.ndarray:
         sc = sc_real + 1j * sc_imag
         return sc.T
 
+    def gain_circle(self, target_port: str, gain: float, npoints: int = 181) -> npy.ndarray:
```
Similar to `stability_circle`

Vinc0110 reviewed (approved):
Very good. Thanks for the changes!
